pkg/utils/fields.go

```diff
@@ -0,0 +1,215 @@
+package utils
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+	"unicode"
+
+	. "github.com/moov-io/irs/pkg/config"
+)
+
+var (
+	alphanumericRegex    = regexp.MustCompile(`[^ \w!"#$%&'()*+,-.\\/:;<>=?@\[\]^_{}|~]+`)
+	numericRegex         = regexp.MustCompile(`^[0-9]+$`)
+	yearRegex            = regexp.MustCompile(`((19|20)\d\d)`)
+	emailRegex           = regexp.MustCompile("^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$")
+	minPhoneNumberLength = 10
+)
+
+// parse field with string
+func ParseValue(fields reflect.Value, spec map[string]SpecField, record string) error {
+	for i := 0; i < fields.NumField(); i++ {
+		fieldName := fields.Type().Field(i).Name
+		// skip local variable
+		if !unicode.IsUpper([]rune(fieldName)[0]) {
+			continue
+		}
+
+		field := fields.FieldByName(fieldName)
+		spec, ok := spec[fieldName]
+		if !ok || !field.IsValid() || !field.CanSet() {
+			return ErrValidField
+		}
+
+		if len(record) < spec.Start+spec.Length {
+			return ErrShortRecord
+		}
+
+		data := record[spec.Start : spec.Start+spec.Length]
+		if err := isValidType(spec, data); err != nil {
+			return err
+		}
+
+		if err := parseValue(spec, field, data); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// to string from field
+func ToString(elm SpecField, data reflect.Value) string {
+	if !data.IsValid() {
+		return fillString(elm)
+	}
+
+	sizeStr := strconv.Itoa(elm.Length)
+	switch elm.Type {
+	case Alphanumeric, Email, Numeric, TelephoneNumber:
+		return fmt.Sprintf("%-"+sizeStr+"s", data)
+	case ZeroNumeric:
+		return fmt.Sprintf("%0"+sizeStr+"d", data)
+	case DateYear:
+		return fmt.Sprintf("%-"+sizeStr+"d", data)
+	}
+
+	return fillString(elm)
+}
+
+// to validate fields of record
+func Validate(r interface{}, spec map[string]SpecField) error {
+	fields := reflect.ValueOf(r).Elem()
+	for i := 0; i < fields.NumField(); i++ {
+		fieldName := fields.Type().Field(i).Name
+		if !fields.IsValid() {
+			return ErrValidField
+		}
+
+		if spec, ok := spec[fieldName]; ok {
+			if spec.Required == Required {
+				fieldValue := fields.FieldByName(fieldName)
+				if fieldValue.IsZero() {
+					return ErrFieldRequired
+				}
+			}
+		}
+
+		funcName := validateFuncName(fieldName)
+		method := reflect.ValueOf(r).MethodByName(funcName)
+		if method.IsValid() {
+			response := method.Call(nil)
+			if len(response) == 0 {
+				continue
+			}
+
+			err := method.Call(nil)[0]
+			if !err.IsNil() {
+				return err.Interface().(error)
+			}
+		}
+	}
+
+	return nil
+}
+
+// File Read
+func ReadFile(f *os.File) []byte {
```

Comment:
Can we use `ioutil.ReadAll(f)` instead? Or `ioutil.ReadFile(path)`. 

https://golang.org/pkg/io/ioutil/#ReadAll

Reply:
fixed